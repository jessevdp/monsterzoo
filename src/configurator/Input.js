import { Component } from '@local/system';

const defaultAttributes = {
    'type': 'text'
};

export default class Input extends Component {
    /**
     * Creates an instance of Input.
     * @param {string} name
     * @param {object} [attributes={}] A set of attributes for the input box, represented as key-value pairs
     * @memberof Input
     */
    constructor(name, attributes = {}) {
        super();
        attributes = {...defaultAttributes, ...attributes};
        this.setState({
            name,
            attributes
        });
    }

    /**
     * Convert the attributes object (stored in state) into it's HTML string
     * equivalent. (E.g. a sequence of: 'key="value"')
     *
     * @returns {string}
     * @memberof Input
     */
    htmlAttributes() {
        return Object.entries(this.state.attributes)
            .map((entry) => `${entry[0]}="${entry[1]}"`)
            .reduce((all, attribute) => all + (attribute + ' '), '')
            .trim();
    }
}