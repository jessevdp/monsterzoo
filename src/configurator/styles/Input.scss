@import '~@local/styles/variables/spacing';

.input {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    > input {
        appearance: none;
        background: transparent;
        border: 0;
        box-sizing: border-box;
        margin: 0;
        padding: 0;

        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);

        &:focus,
        &:active {
            outline: 0;
        }
    }
}