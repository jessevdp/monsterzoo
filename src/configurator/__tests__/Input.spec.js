import Input from '../Input';

it('renders correctly', () => {
    const input = new Input('foobar');
    input.state.attributes = { foo: 'bar', bar: 'foo' };
    expect(input.view()).toMatchSnapshot();
    input.cleanup();
})

describe('constructor', () => {
    it('sets the [name] on state', () => {
        const name = 'foobar';
        const input = new Input(name);
        expect(input.state.name).toBe(name);
        input.cleanup();
    })
    it('uses defaults for the [attributes]', () => {
        const input = new Input('foobar');
        expect(input.state.attributes).toMatchSnapshot();
        input.cleanup();
    })
    it('merges the passed attributes with the defaults', () => {
        const input = new Input('foobar', { foo: 'bar' });
        expect(input.state.attributes).toHaveProperty('foo', 'bar');
        input.cleanup();
    })
    it('overwrites attribute defaults with passed attributes', () => {
        const input = new Input('foobar', { type: 'number' });
        expect(input.state.attributes.type).toBe('number');
        input.cleanup();
    })
})

describe('htmlAttributes', () => {
    it('can convert [attributes] into an HTML string', () => {
        const attributes = { foo: 'bar', number: 10 }
        const expected = 'foo="bar" number="10"';
        const input = new Input('name');
        input.state.attributes = attributes;
        expect(input.htmlAttributes()).toBe(expected);
        input.cleanup();
    })
})

describe('setAttributes', () => {
    it('accepts an object to update the attributes', () => {
        const input = new Input('name');
        const attributes = { foo: 'bar' };
        input.setAttributes(attributes);
        expect(input.state.attributes.foo).toBe(attributes.foo);
        input.cleanup();
    })
    it('accepts a function to update the attributes', () => {
        const input = new Input('name');
        const attributes = { foo: 'bar' };
        input.setAttributes(() => attributes);
        expect(input.state.attributes.foo).toBe(attributes.foo);
        input.cleanup();
    })
    it('passes the old attributes to the passed function', () => {
        // Arrange
        const oldAttributes = { foo: 'foo' };
        const newAttributes = { foo: 'bar' };
        const fn = jest.fn(() => newAttributes);
        const input = new Input('name');
        input.state.attributes = oldAttributes;

        // Act
        input.setAttributes(fn);

        // Assert
        expect(fn).toHaveBeenCalledWith(oldAttributes);

        // Cleanup
        input.cleanup();
    })
    it('overwrites old values', () => {
        const oldAttributes = { foo: 'foo' };
        const newAttributes = { foo: 'bar' };
        const input = new Input('name', oldAttributes);
        input.setAttributes(newAttributes);
        expect(input.state.attributes.foo).toBe(newAttributes.foo);
        input.cleanup();
    })
    it('leaves other values intact', () => {
        const input = new Input('name');
        const fooAttribute = { foo: 'foo' };
        const barAttribute = { bar: 'bar' };
        input.setAttributes(fooAttribute);
        input.setAttributes(barAttribute);
        expect(input.state.attributes.foo).toBe(fooAttribute.foo);
        expect(input.state.attributes.bar).toBe(barAttribute.bar);
        input.cleanup();
    })
})