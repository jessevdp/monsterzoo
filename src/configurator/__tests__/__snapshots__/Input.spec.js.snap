// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor uses defaults for the [attributes] 1`] = `
Object {
  "type": "text",
}
`;
